src/lightgbm/src/main/scala/LightGBMRegressor.scala

```diff
@@ -71,7 +71,7 @@ class LightGBMRegressor(override val uid: String)
     log.info(s"Nodes used for LightGBM: ${nodes.mkString(",")}")
     val trainParams = RegressorTrainParams(getParallelism, getNumIterations, getLearningRate, getNumLeaves,
       getApplication, getAlpha, getMaxBin, getBaggingFraction, getBaggingFreq, getBaggingSeed, getFeatureFraction,
-      getMaxDepth, getMinSumHessianInLeaf)
+      getMaxDepth, getMinSumHessianInLeaf, nodes.length)
```

Comment:
I run this new version in our environment yesterday and I did see only slight improvement and I think we suffer from the same problem (the iterations are running out of sync), as our results in spark are not as good as the results we are getting on the same data in R.

@ywskycn @imatiach-msft i think here we should have number of partitions not the number of machines (in our case the number of machines is 3 and number of partitions 96) and i see in the logs:

```
18/06/19 20:15:37 INFO LightGBMRegressor: Nodes used for LightGBM: (1,172.33.8.230),(2,172.33.0.91),(0,172.33.5.247)
```

And I still the see the lines like these below that IMHO suggest that we are not syncing the model properly.

```

June 19th 2018, 22:22:11.745 | 18/06/19 20:22:10 INFO LightGBMRegressor: LightGBM running iteration: 5935 with result: 0 and is finished: 0
-- | --

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:10 INFO LightGBMRegressor: LightGBM running iteration: 5989 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:10 INFO LightGBMRegressor: LightGBM running iteration: 5889 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:10 INFO LightGBMRegressor: LightGBM running iteration: 5999 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:11 INFO LightGBMRegressor: LightGBM running iteration: 5978 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:10 INFO LightGBMRegressor: LightGBM running iteration: 5933 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:11 INFO LightGBMRegressor: LightGBM running iteration: 5934 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:11 INFO LightGBMRegressor: LightGBM running iteration: 5991 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:11 INFO LightGBMRegressor: LightGBM running iteration: 5939 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:10 INFO LightGBMRegressor: LightGBM running iteration: 5887 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:10 INFO LightGBMRegressor: LightGBM running iteration: 5975 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:11 INFO LightGBMRegressor: LightGBM running iteration: 5937 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:11 INFO LightGBMRegressor: LightGBM running iteration: 5976 with result: 0 and is finished: 0

  | June 19th 2018, 22:22:11.745 | 18/06/19 20:22:10 INFO LightGBMRegressor: LightGBM running iteration: 5999 with result: 0 and is finished: 0

```

Reply:
I think here the nodes refers to the partitions, from https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMUtils.scala#L189.

If the input DataFrame's partitions is larger than #executors * #cores, it should be coalesced at the beginning. https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMRegressor.scala#L62.

Do I miss anything here?

Reply:
@ywskycn i think the nodes refers to the spark executors - as it gets it from this method:
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMUtils.scala#L145

and in my case in line https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMRegressor.scala#L71
it prints:

```
18/06/19 20:15:37 INFO LightGBMRegressor: Nodes used for LightGBM: (1,172.33.8.230),(2,172.33.0.91),(0,172.33.5.247)
```

While later in the process I am having the following debug from LightGBM:

```
18/06/20 12:40:58 INFO LightGBMRegressor: driver writing back to all connections: 172.33.4.88:12410,172.33.4.88:12419,172.33.4.88:12411,172.33.4.88:12417,172.33.4.88:12412,172.33.4.88:12403,172.33.4.88:12409,172.33.4.88:12401,172.33.4.88:12415,172.33.4.88:12402,172.33.4.88:12407,172.33.4.88:12405,172.33.4.88:12418,172.33.4.88:12408,172.33.4.88:12416,172.33.4.88:12404,172.33.4.88:12406,172.33.4.88:12400,172.33.4.88:12413,172.33.4.88:12414
```

which is displaying all the LightGBM workers used in the training and it reflects the number of partitions. 

(Those two  lines are from different trainings so the IPs can differ).

Reply:
@troszok  @ywskycn  this is a very good catch, that's actually the number of executors, not nodes

Reply:
@troszok @ywskycn I think you want to pass in numExecutorCores